frameaccurate/js/FrameAccurateControls.ts

```diff
@@ -0,0 +1,255 @@
+      const debugSmpte = smpte.toString();
+      const targetTime = smpte.toAdjustedTime();
+      console.debug('Seeking to SMTPE: ' + debugSmpte + ', calculated Time: ' + targetTime);
+      this.player.seek(targetTime);
+    }
+    catch (error) {
+      console.error('Error during converting smtpe to time: ' + error);
+      throw error;
+    }
+  }
+
+  public getCurrentSmpte(): string {
+    const currentTime = this.player.getCurrentTime();
+    return SmpteTimestamp.fromTimeWithAdjustments(currentTime, this.assetDescription).toString();
+  }
+
+  public step(stepSize: number) {
+    const smpte = new SmpteTimestamp(this.getCurrentSmpte(), this.assetDescription);
+    if(smpte.minutes % 10 !== 0 && smpte.seconds === 0) {
+      //  in the case of being around the dropped frame at step start we have to ignore the frameHoles
+      smpte.addFrame(stepSize, false);
+    } else {
+      smpte.addFrame(stepSize, true);
+    }
+    this.seekToSMPTE(smpte.toString());
+  }
+}
+
+class AssetDescription {
+  public frameDuration: number;
+  public offsetToMidFrame: number;
+  constructor(public name: string, public sourceConfig: SourceConfig, public framesPerSecond: number,
+              public adjustmentFactor?: number, public framesDroppedAtFullMinute?: number) {
+    if (adjustmentFactor == null) {
+      this.adjustmentFactor = Math.ceil(framesPerSecond) / framesPerSecond;
+    }
+    if(framesDroppedAtFullMinute == null) {
+      if (framesPerSecond === 29.97 || framesPerSecond === 29.98) {
+        this.framesDroppedAtFullMinute = 2;
+      }
+      else {
+        this.framesDroppedAtFullMinute = 0;
+      }
+    }
+    this.frameDuration = 1 / this.framesPerSecond;
+    // needed to not seek to the end of the previous frame but rather to the middle of the desired one
+    this.offsetToMidFrame = (this.frameDuration / 2) / this.adjustmentFactor;
+  }
+}
+
+interface SourceConfig {
+  dash?: string;
+  hls?: string;
+  progressive?: string;
+  smooth?: string;
+}
+
+class SmpteTimestamp {
+  public frame: number;
+  public seconds: number;
+  public minutes: number;
+  public hours: number;
+
+  public constructor(smtpeTimestamp: string, private assetDescription: AssetDescription) {
+    if (smtpeTimestamp && SmpteTimestamp.validateTimeStamp(smtpeTimestamp, assetDescription.framesPerSecond)) {
+
+      const parts: string[] = smtpeTimestamp.split(':');
+
+      this.hours = Number(parts[0]);
+      this.minutes = Number(parts[1]);
+      this.seconds = Number(parts[2]);
+      this.frame = Number(parts[3]);
+    } else {
+      this.hours = 0;
+      this.minutes = 0;
+      this.seconds = 0;
+      this.frame = 0;
+    }
+  }
+
+  public static validateTimeStamp(smtpeTimestamp: string, framesPerSecond: number): boolean {
+
+    // validate SMTPE timecode
+    let isValidSMPTETimeCode = new RegExp(/(^(?:(?:[0-1][0-9]|[0-2][0-3]):)(?:[0-5][0-9]:){2}(?:[0-6][0-9])$)/);
```

Comment:
From [Wikipedia](https://en.wikipedia.org/wiki/SMPTE_timecode#Drop_frame_timecode): 
> While non-drop timecode is displayed with colons separating the digit pairs—"HH:MM:SS:FF"—drop frame is usually represented with a semi-colon (;) or period (.) as the divider between all the digit pairs—"HH;MM;SS;FF", "HH.MM.SS.FF"—or just between the seconds and frames—"HH:MM:SS;FF" or "HH:MM:SS.FF". The period is usually used on VTRs and other devices that don't have the ability to display a semi-colon.
>
> Drop frame timecode is typically abbreviated as DF and non-drop as NDF.

Is this taken into account? We should use the `;` only between seconds and frames for this case.

Reply:
This is not taken into account, but would be not be too hard to implement